include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+                }
+        }
+
+        return offsetToTextureData(pgtOffset, _img, _wrapu, _wrapv);
+    }
+
+    bool free(const STextureData& _addr, const IImage* _img, const IImage::SSubresourceRange& _subres) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
+        ICPUVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return false;
+            storage = static_cast<ICPUVTResidentStorage*>(found->second.get());
+        }
+
+        //free physical pages
+        auto extent = _img->getCreationParameters().extent;
+        const uint32_t levelCount = countLevelsTakingAtLeastOnePage(extent, _subres);
+
+        CFillImageFilter::state_type fill;
+        fill.outImage = m_pageTable.get();
+        fill.subresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+        fill.subresource.baseArrayLayer = _addr.pgTab_layer;
+        fill.subresource.layerCount = 1u;
+        fill.fillValue.asUint.x = SPhysPgOffset::invalid_addr;
+
+        using phys_pg_addr_alctr_t = ICPUVTResidentStorage::phys_pg_addr_alctr_t;
+
+        auto* const bufptr = reinterpret_cast<uint8_t*>(m_pageTable->getBuffer()->getPointer());
+        for (uint32_t i = 0u; i < levelCount; ++i)
+        {
+            const uint32_t w = neededPageCountForSide(extent.width, _subres.baseMipLevel+i);
+            const uint32_t h = neededPageCountForSide(extent.height, _subres.baseMipLevel+i);
+
+            const auto& region = m_pageTable->getRegions().begin()[i];
+            const auto strides = region.getByteStrides(TexelBlockInfo(m_pageTable->getCreationParameters().format));
+
+            for (uint32_t y = 0u; y < h; ++y)
+                for (uint32_t x = 0u; x < w; ++x)
+                {
+                    uint32_t* texelptr = reinterpret_cast<uint32_t*>(bufptr + region.getByteOffset(core::vector4du32_SIMD((_addr.pgTab_x>>i) + x, (_addr.pgTab_y>>i) + y, 0u, _addr.pgTab_layer), strides));
+                    SPhysPgOffset physPgOffset = *texelptr;
+                    if (storage->physPgOffset_valid(physPgOffset))
+                    {
+                        *texelptr = SPhysPgOffset::invalid_addr;
+
+                        uint32_t addrs[2] { physPgOffset.addr&0xffffu, storage->physPgOffset_mipTailAddr(physPgOffset).addr&0xffffu };
+                        const uint32_t szs[2]{ 1u,1u };
+                        core::address_allocator_traits<phys_pg_addr_alctr_t>::multi_free_addr(storage->tileAlctr, storage->physPgOffset_hasMipTailAddr(physPgOffset) ? 2u : 1u, addrs, szs);
+                    }
+                }
+            fill.subresource.mipLevel = i;
+            fill.outRange.offset = {static_cast<uint32_t>(_addr.pgTab_x>>i),static_cast<uint32_t>(_addr.pgTab_y>>i),0u};
+            fill.outRange.extent = {w,h,1u};
+            CFillImageFilter::execute(&fill);
```

Comment:
See https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension&hide-deleted-files=true#r417905535

Reply:
plus anyway, you're doing it twice... cause you've already done it here?
https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension&hide-deleted-files=true#r417913790